cirq/ops/three_qubit_gates.py

```diff
@@ -145,6 +144,125 @@ def __str__(self) -> str:
+        return protocols.CircuitDiagramInfo(('diag', '#2', '#3'))
+
+    def __pow__(self, exponent: Any) -> 'ThreeQubitDiagonalGate':
+        if not isinstance(exponent, (int, float)):
+            return NotImplemented
+        return ThreeQubitDiagonalGate(
+            *[angle * exponent for angle in self._diag_angles])
+
+    def _decompose_(self, qubits):
+        """An adjacency-respecting decomposition.
+
+        0: ───p_0───@──────────────@───────@──────────@──────────
+                    │              │       │          │
+        1: ───p_1───X───@───p_3────X───@───X──────@───X──────@───
+                        │              │          │          │
+        2: ───p_2───────X───p_4────────X───p_5────X───p_6────X───
+
+        where p_i = T**(4*x_i) and x_i solve the system of equations
+                    [0, 0, 1, 0, 1, 1, 1][x_0]   [l_1]
+                    [0, 1, 0, 1, 1, 0, 1][x_1]   [l_2]
+                    [0, 1, 1, 1, 0, 1, 0][x_2]   [l_3]
+                    [1, 0, 0, 1, 1, 1, 0][x_3] = [l_4]
+                    [1, 0, 1, 1, 0, 0, 1][x_4]   [l_5]
+                    [1, 1, 0, 0, 0, 1, 1][x_5]   [l_6]
+                    [1, 1, 1, 0, 1, 0, 0][x_6]   [l_7]
+        where l_i is self._diag_angles[i].
+
+        The above system was created by equating the composition of the gates
+        in the circuit diagram to np.diag(self._diag_angles) (shifted by a
+        global phase of np.exp(-1j * self._diag_angles[0])).
+        """
+
+        a, b, c = qubits
+        if hasattr(b, 'is_adjacent'):
+            if not b.is_adjacent(a):
+                b, c = c, b
+            elif not b.is_adjacent(c):
+                a, b = b, a
+        sweep_abc = [common_gates.CNOT(a, b), common_gates.CNOT(b, c)]
+        phase_matrix = np.array([[0, 0, 1, 0, 1, 1, 1], [0, 1, 0, 1, 1, 0, 1],
+                                 [0, 1, 1, 1, 0, 1, 0], [1, 0, 0, 1, 1, 1, 0],
+                                 [1, 0, 1, 1, 0, 0, 1], [1, 1, 0, 0, 0, 1, 1],
+                                 [1, 1, 1, 0, 1, 0, 0]])
+        shifted_angles_tail = [
+            angle - self._diag_angles[0] for angle in self._diag_angles[1:]
+        ]
+        phase_solutions = np.linalg.solve(phase_matrix, shifted_angles_tail)
+        p_gates = []
+        for solution in phase_solutions:
+            p_gates.append(common_gates.T**(4 * solution / np.pi))
+
+        return [
+            p_gates[0](a),
+            p_gates[1](b),
+            p_gates[2](c),
+            sweep_abc,
+            p_gates[3](b),
+            p_gates[4](c),
+            sweep_abc,
+            p_gates[5](c),
+            sweep_abc,
+            p_gates[6](c),
+            sweep_abc,
+        ]
+
+    def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs) -> np.ndarray:
+        if np.allclose(self._matrix, np.identity(8)):
+            return args.target_tensor
+        return linalg.targeted_left_multiply(self._matrix.astype(
```

Comment:
This can be implemented much more efficiently by taking advantage of the fact that the matrix is diagonal. Iterate over the diagonal elements while phasing the corresponding subsection of the tensor. There is a utility method `subspace_index` on the `args` value to make it easier to get at the relevant subsection of the tensor.

Reply:
Let me know if the subspace indexing is too hacky.